internal/docker/deployer.go

```diff
@@ -212,12 +219,14 @@ func deployImage(
 		PortBindings: nat.PortMap{
 			nat.Port("8008/tcp"): []nat.PortBinding{
 				{
-					HostIP: "127.0.0.1",
+					HostIP:   "127.0.0.1",
+					HostPort: strconv.Itoa(port1),
 				},
 			},
 			nat.Port("8448/tcp"): []nat.PortBinding{
 				{
-					HostIP: "127.0.0.1",
+					HostIP:   "127.0.0.1",
+					HostPort: strconv.Itoa(port2),
```

Comment:
We must provide an explicit port, otherwise we'll get a different random port when the container is restarted. Which breaks everything.

Reply:
`BaseURL` and `FedBaseURL` of `HomeserverDeployment` become incorrect because they include the port number. We can fix up those URLs. But any `CSAPI` client instances will continue to use the previous port number and that's not so easy to fix up.

Reply:
That seems like a reasonably easy thing to fix? There's several possible solutions:
 - Add documentation to `Restart()` which explicitly states that any clients created with this deployment need to be recreated. Or...
 - Add a callback hook to `client.CSAPI` to update the base URL and get test authors to call that callback with the updated URL. Or...
 - Remember all `client.CSAPI` instances created via `func (d *Deployment) Client(t *testing.T, hsName, userID string) *client.CSAPI` and automatically call the callback hook when the deployment is restarted to re-point port numbers.

The last option is the most preferable because:
 - All the complexity is in one place and does not need to be repeated in tests.
 - Test authors don't need to know or care that the port number changes from under them when they call `deployment.Restart()`.

In addition, I would change the function signature of restart from:
```go
func (dep *Deployment) Restart() error
```
to
```go
func (dep *Deployment) Restart(t *testing.T)
```
which will fail the test if the restart fails (returns an error).

At present, this PR cannot be accepted because it will introduce flakiness because these port numbers will race: we have to let the underlying OS decide the ports.

Reply:
I've split the PR and made the changes in #396. Let me know if I've missed anything.